### PR TITLE
[NVIDIA XLA GPU] Restore dotdimension numbers for original bmms

### PR DESCRIPTION
**Pull request by Tixxx**
This is trying to fix the error seen in https://github.com/openxla/xla/pull/5782.
This was supposed to be in one of the previous bug fixes but got lost when porting from internal PR to github.

### CODE REVIEW

**Review by cheshire — CHANGES_REQUESTED** [review]
Is it possible to add a test?

**Comment by akuegel** [comment]
Here is the HLO with the minimized reproducer which resulted in the rollback:

```
HloModule test

%region_50.2457 (Arg_0.2458: bf16[], Arg_1.2459: bf16[]) -> bf16[] {
  %Arg_0.2458 = bf16[] parameter(0)
  %Arg_1.2459 = bf16[] parameter(1)
  ROOT %maximum.2 = bf16[] maximum(bf16[] %Arg_0.2458, bf16[] %Arg_1.2459)
}

%region_36.2316 (Arg_0.2317: f32[], Arg_1.2318: f32[]) -> f32[] {
  %Arg_0.2317 = f32[] parameter(0)
  %Arg_1.2318 = f32[] parameter(1)
  ROOT %add.342 = f32[] add(f32[] %Arg_0.2317, f32[] %Arg_1.2318)
}

ENTRY main {
  %transpose.482 = bf16[4,5,64]{2,1,0} parameter(0)
  %transpose.484 = bf16[4,64,5]{2,1,0} parameter(1)
  %dot.20 = bf16[4,5,5]{2,1,0} dot(bf16[4,5,64]{2,1,0} %transpose.482, bf16[4,64,5]{2,1,0} %transpose.484), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={1}
  %constant.2515 = bf16[] constant(0.125)
  %broadcast.789 = bf16[4,5,5]{2,1,0} broadcast(bf16[] %constant.2515), dimensions={}
  %multiply.267 = bf16[4,5,5]{2,1,0} multiply(bf16[4,5,5]{2,1,0} %dot.20, bf16[4,5,5]{2,1,0} %broadcast.789)
  %constant.287 = f32[] constant(-1)
  %broadcast.792 = bf16[4,5,5]{2,1,0} parameter(3)
  %add.348 = bf16[4,5,5]{2,1,0} add(bf16[4,5,5]{2,1,0} %multiply.267, bf16[4,5,5]{2,1,0} %broadcast.792)
  %constant.2510 = bf16[] constant(-inf)
  %reduce.2550 = bf16[4,5]{1,0} reduce(bf16[4,5,5]{2,1,0} %add.348, bf16[] %constant.2510), dimensions={2}, to_apply=%region_50.2457
  %broadcast.793 = bf16[4,5,5]{2,1,0} broadcast(bf16[4,5]{1,0} %reduce.2550), dimensions={0,1}
  %subtract.81 = bf16[4,5,5]{2,1,0} subtract(bf16[4,5,5]{2,1,0} %add.348, bf16[4,5,5]{2,1,0} %broadcast.793)
  %exponential.21 = bf16[4,5,5]{2,1,0} exponential(bf16[4,5,5]{2,1,0} %subtract.81)
  %convert.180 = f32[4,5,5]{2,1,0} convert(bf16[4,5,5]{2,1,0} %exponential.21)
  %constant.2509 = f32[] constant(0)
  %reduce.2558 = f32[4,5]{1,0} reduce(f32[4,5,5]{2,1,0} %convert.180, f32[] %constant.2509), dimensions={2}, to_apply=%region_36.2316
  %convert.182 = bf16[4,5]{1,0} convert(f32[4,5]{1,0} %reduce.2558)
  %broadcast.794 = bf16[4,5,5]{2,1,0} broadcast(bf16[4,5]{1,0} %convert.182), dimensions={0,1}
  %divide.25 = bf16[4,5,5]{2,1,0} divide(bf16[4,5,5]{2,1,0} %exponential.21, bf16[4,5,5]{2,1,0} %broadcast.794)
  %transpose.481 = bf16[4,64,5]{2,1,0} parameter(2)
  %dot.21 = bf16[4,64,5]{2,1,0} dot(bf16[4,64,5]{2,1,0} %transpose.481, bf16[4,5,5]{2,1,0} %divide.25), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={2}
  ROOT %tuple.2668 = (bf16[4,5,5]{2,1,0}, bf16[4,64,5]{2,1,0}) tuple(bf16[4,5,5]{2,1,0} %divide.25, bf16[4,64,5]{2,1,0} %dot.21)
}
```

I guess you can create a test out of that :)

**Review by akuegel — COMMENTED** [review]
What about the calls to ChangeCheckedDimToFastest in FusedBwdMultiHeadedAttentionBlock? I would assume it is also necessary to restore the dot dimension numbers there.

Also, what I mentioned in the other PR is that I am not sure whether it is a good idea to match cases where the dot will be run again (because there is a direct or indirect user outside of the matched pattern). Have you run any benchmarks to check this?
So personally I would assume it is only a good idea to match if every part of what is matched into the custom call has no outside user, except the root operation.

**Comment by Tixxx** [comment]
> Is it possible to add a test?

Added more tests to exercise more negative cases.

**Comment by Tixxx** [comment]
> What about the calls to ChangeCheckedDimToFastest in FusedBwdMultiHeadedAttentionBlock? I would assume it is also necessary to restore the dot dimension numbers there.
> 
> Also, what I mentioned in the other PR is that I am not sure whether it is a good idea to match cases where the dot will be run again (because there is a direct or indirect user outside of the matched pattern). Have you run any benchmarks to check this? So personally I would assume it is only a good idea to match if every part of what is matched into the custom call has no outside user, except the root operation.

Yea, I added the logic to restore all bmms in cases of failure to match.

As for the second point, we do a progressive lowering where we lower forward, match backward and then lower backward. This means if forward matches but backward fails to match, the forward graph has been already modified. I pushed a new commit to restore the forward graph structure and remove the forward FMHA call if backward fails to match. If forward fails to match directly, we wont have this type of issues.

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 1]
```diff
@@ -1288,9 +1296,15 @@ StatusOr<HloInstruction*> FuseFwdMultiHeadedAttentionBlock(
...
-    TF_RETURN_IF_ERROR(comp->ReplaceWithNewInstruction(
-        activation_output, HloInstruction::CreateGetTupleElement(
-                               activation_output->shape(), fmha_call, 2)));
+    HloInstruction* activation_gte =
+        comp->AddInstruction(HloInstruction::CreateGetTupleElement(
+            activation_output->shape(), fmha_call, 2));
+    TF_ASSIGN_OR_RETURN(
+        bool activation_changed,
+        comp->ReplaceInstructionWithDifferentShape(
+            activation_output, activation_gte, /*preserve_sharding*/ false,
```
[the remark this answers is not in the document]
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc]: updated

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 2]
```diff
@@ -1544,6 +1576,28 @@ StatusOr<bool> CudnnFusedMHARewriter::Run(
...
               matched_result.matched_custom_call_name, debug_options));
       if (!is_mha_module_supported) continue;
+
+      // If we have an activation with more than 1 users in non-training mode,
+      // we cannot rewrite the graph. So skip processing the rest.
+      HloInstruction* activation =
+          matched_result.need_canonicalization
+              ? matched_result.matched_bmm_2->mutable_operand(1)
+              : matched_result.matched_bmm_2->mutable_operand(0);
+      if (!matched_result.is_training && activation->user_count() > 1) {
```
[the remark this answers is not in the document]
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:1586]: Yes, we are working on that too. I will be OOO starting tomorrow for 2 weeks. @Cjkkkk from our team will take over the PR and possibly adding additional checks.

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 3]
```diff
@@ -671,11 +671,10 @@ bool IsBmm2GradGemm2(HloInstruction* instr) {
 }
 
 MatchBwdResult MatchBmm1GradGemm1(MatchBwdResult previous_result,
-                                  HloInstruction* fwd_fmha_call,
                                   HloInstruction* bmm_1) {
   MatchBwdResult match_result = previous_result;
   match_result.has_match = false;
-  const HloInstruction* q_tensor = fwd_fmha_call->operand(0);
+  const HloInstruction* q_tensor = bmm_1->operand(0);
```
Comment (akuegel) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:677]: Another test that covers this change would be nice :)

**Comment by Cjkkkk** [comment]
//xla/hlo/experimental/auto_sharding:auto_sharding_test                  FAILED in 3 out of 3 in 2.3s.
Do we expect this CPU test to fail or should I investigate this?

**Comment by akuegel** [comment]
> //xla/hlo/experimental/auto_sharding:auto_sharding_test FAILED in 3 out of 3 in 2.3s. Do we expect this CPU test to fail or should I investigate this?

This test is currently expected to fail in open source testing, so no need to investigate.

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 3 above)
Comment (Cjkkkk) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:677]: Add one test case where Q is transposed because head dim does not have stride 1(layout constraint required by cuDNN backend). Improved bwd pattern match will pattern match in this case.

**Comment by Cjkkkk** [comment]
@akuegel, Hi do we have anything pending for this PR?

**Review by akuegel — APPROVED** [review]

**Comment by AyanmoI** [comment]
@akuegel @Cjkkkk This is a followup to turning on the fmha feature by default. Can we add the turning on by default as part of this PR?

**Comment by akuegel** [comment]
I think it would be better to wait until also the matching is improved to not match if there are outside users. Which should be another PR

**Comment by AyanmoI** [comment]
> I think it would be better to wait until also the matching is improved to not match if there are outside users. Which should be another PR

@Cjkkkk can you create the new PR for this.

**Comment by Cjkkkk** [comment]
@akuegel @AyanmoI I will create a new PR to include additional user check and turning on the fmha feature by default.
